File: src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds/src/publisher.cpp
// Copyright (C) 2022 Intel Corporation
// SPDX-License-Identifier: BSD-3-Clause
// See: https://spdx.org/licenses/

#include <chrono>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "ddsmetadata/msg/dds_meta_data.hpp"

using namespace std::chrono_literals;

#define MAX_ARRAY_DIMS (5)

struct MetaData {
  int64_t nd;
  int64_t type;
  int64_t elsize;
  int64_t total_size;
  int64_t dims[MAX_ARRAY_DIMS] = {0};
  int64_t strides[MAX_ARRAY_DIMS] = {0};
  void* mdata;
};

using MetaDataPtr = std::shared_ptr<MetaData>;

class MinimalPublisher : public rclcpp::Node
{
public:
  MinimalPublisher()
  : Node("minimal_publisher"), count_(0)
  {
    metadata = std::make_shared<MetaData>();
    metadata->nd = 1;
    metadata->type = 7;
    metadata->elsize = 8;
    metadata->total_size = 1;
    metadata->dims[0] = 1;
    metadata->strides[0] = 1;
    metadata->mdata = reinterpret_cast<char*> (malloc(sizeof(int64_t)));
    *reinterpret_cast<int64_t*>(metadata->mdata) = 0;

    publisher_ = this->create_publisher<ddsmetadata::msg::DDSMetaData>("dds_topic", rclcpp::SystemDefaultsQoS());
    timer_ = this->create_wall_timer(
      500ms, std::bind(&MinimalPublisher::timer_callback, this));
  }

private:
  void timer_callback()
  {
    auto message = ddsmetadata::msg::DDSMetaData();
    message.nd = metadata->nd;
    message.type = metadata->type;
    message.elsize = metadata->elsize;
    message.total_size = metadata->total_size;
    for(int i = 0; i < MAX_ARRAY_DIMS; i++) {
      message.dims[i] = metadata->dims[i];
      message.strides[i] = metadata->strides[i];
    }
    *reinterpret_cast<int64_t*>(metadata->mdata) = count_;
    message.mdata = std::vector<unsigned char>(
                  reinterpret_cast<unsigned char*>(metadata->mdata),
                  reinterpret_cast<unsigned char*>(metadata->mdata) + metadata->elsize * metadata->total_size);
    
    RCLCPP_INFO(this->get_logger(), "ROS2 publishing: '%d'", count_++);
    publisher_->publish(message);
  }
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<ddsmetadata::msg::DDSMetaData>::SharedPtr publisher_;
  size_t count_;
  MetaDataPtr metadata;
  rmw_publisher_t publisher;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MinimalPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/publisher.py
# Copyright (C) 2022 Intel Corporation
# SPDX-License-Identifier: BSD-3-Clause
# See: https://spdx.org/licenses/
import numpy as np
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_system_default

from ddsmetadata.msg import DDSMetaData
from .utils.np_mdata_trans import nparray_to_metadata


class MinimalPublisher(Node):

    def __init__(self):
        super().__init__('minimal_publisher')
        self.publisher_ = self.create_publisher(DDSMetaData,
                                                'dds_topic',
                                                qos_profile_system_default)
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.i = 0

    def timer_callback(self):
        np_arr = np.array(([self.i, 2, 3]), np.int64)
        msg = nparray_to_metadata(np_arr)
        self.publisher_.publish(msg)
        print("Publishing : ", np_arr)
        self.i += 1


def main(args=None):
    rclpy.init(args=args)

    minimal_publisher = MinimalPublisher()

    rclpy.spin(minimal_publisher)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    minimal_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
